lib/Builtins.cpp

```diff
@@ -14,171 +14,339 @@
-  return name.startswith("_Z16get_image_height14ocl_image2d_ro") ||
-         name.startswith("_Z16get_image_height14ocl_image2d_wo") ||
-         name.startswith("_Z16get_image_height14ocl_image3d_ro") ||
-         name.startswith("_Z16get_image_height14ocl_image3d_wo") ||
-         name.startswith("_Z16get_image_height20ocl_image2d_array_ro") ||
-         name.startswith("_Z16get_image_height20ocl_image2d_array_wo");
-}
-
-bool clspv::IsGetImageWidth(StringRef name) {
-  return name.startswith("_Z15get_image_width14ocl_image1d_ro") ||
-         name.startswith("_Z15get_image_width14ocl_image1d_wo") ||
-         name.startswith("_Z15get_image_width14ocl_image2d_ro") ||
-         name.startswith("_Z15get_image_width14ocl_image2d_wo") ||
-         name.startswith("_Z15get_image_width14ocl_image3d_ro") ||
-         name.startswith("_Z15get_image_width14ocl_image3d_wo") ||
-         name.startswith("_Z15get_image_width20ocl_image1d_array_ro") ||
-         name.startswith("_Z15get_image_width20ocl_image1d_array_wo") ||
-         name.startswith("_Z15get_image_width20ocl_image2d_array_ro") ||
-         name.startswith("_Z15get_image_width20ocl_image2d_array_wo");
-}
+    { "__clspv_vloada_half2",                      Builtins::EVloadaHalf                             },
+    { "__clspv_vloada_half4",                      Builtins::EVloadaHalf                             },
+ 
+  };
 
-bool clspv::IsGetImageDepth(StringRef name) {
-  return name.startswith("_Z15get_image_depth14ocl_image3d_ro") ||
-         name.startswith("_Z15get_image_depth14ocl_image3d_wo");
-}
-
-bool clspv::IsGetImageDim(StringRef name) {
-  return name.startswith("_Z13get_image_dim14ocl_image2d_ro") ||
-         name.startswith("_Z13get_image_dim14ocl_image2d_wo") ||
-         name.startswith("_Z13get_image_dim14ocl_image3d_ro") ||
-         name.startswith("_Z13get_image_dim14ocl_image3d_wo") ||
-         name.startswith("_Z13get_image_dim20ocl_image2d_array_ro") ||
-         name.startswith("_Z13get_image_dim20ocl_image2d_array_wo");
+  auto ii = s_func_map.find(_name.c_str());
+  if (ii != s_func_map.end()) {
+    return (*ii).second;
+  }
+  return Builtins::EBuiltinNone;
 }
 
-bool clspv::IsImageQuery(StringRef name) {
-  return clspv::IsGetImageHeight(name) || clspv::IsGetImageWidth(name) ||
-         clspv::IsGetImageDepth(name) || clspv::IsGetImageDim(name);
+////////////////////////////////////////////////////////////////////////////////
+////  FunctionInfo
+////////////////////////////////////////////////////////////////////////////////
+
+static bool GetUnmangledName(const std::string& str, size_t& pos, std::string& name) {
+  char* end;
+  int nameLen = strtol(&str[pos], &end, 10);
+  if (!nameLen) {
+    return false;
+  }
+  pos = end - str.data();
+
+  name = str.substr(pos, nameLen);
+  pos += nameLen;
+  return true;
 }
+
+static bool GetParameterType(clspv::Builtins::ParamTypeInfo& ti, const std::string& name, size_t& pos) {
```

Comment:
Reference parameters should be const. 
(Except in very rare circumstances).  https://google.github.io/styleguide/cppguide.html#Reference_Arguments

At the call site, it's not obvious that pos is being modified.
I don't know about "ti" either, and at this point of reading the text, it's not clear.